### PR TITLE
STL_extension: Introduce CGAL::iterator 

### LINKED ISSUE #2832: std::iterator depreciation (c++17)
## Issue Details

`std::iterator<> ` is deprecated with C++17 with no replacement (replacement is "write the typedefs in the class") and is used in a few packages, mainly in `CGAL/iterator.h`.

`CGAL::unary_function` was introduced, so should `CGAL::iterator` be introduced too ?

### PR DESCRIPTION
**Pull request by afabri**
... as replacement for the deprecated `std::iterator`.  It is only an implementation detail, that is we do not document it. 

## Release Management

* Affected package(s): `STL_extension`, ` Algebraic_kernel_d`, `Arrangement_on_surface_2`, `GraphicsView` 
* Issue(s) solved (if any): fix #2832

### CODE REVIEW

File: STL_Extension/include/CGAL/iterator.h [hunk 1]
```diff
@@ -286,7 +303,7 @@ class Counting_output_iterator
 };
 
 template < class I,
-           class Val = typename std::iterator_traits<I>::value_type >
+           class Val = typename CGAL::iterator_traits<I>::value_type >
```
Comment (MaelRL) [inline STL_Extension/include/CGAL/iterator.h]: Typo

Comment (lrineau) [reply STL_Extension/include/CGAL/iterator.h]: Reminder: https://github.com/CGAL/cgal/wiki/Testing#using-ctest

;-)

**Comment by lrineau** [comment]
Another note, about the first message at https://github.com/CGAL/cgal/pull/2839#issue-169191866:
> ## Release Management
> 
> * Affected package(s): STL_extension
> * Issue(s) solved (if any): fix #2832

`STL_extension` is clearly not the only affected package. If I cannot trust that "Release Management" section, then I (as Release Manager) have to verify more.

File: STL_Extension/include/CGAL/iterator.h (hunk 1 above)
Comment (afabri) [reply STL_Extension/include/CGAL/iterator.h]: Well, ` std::iterator` still exists. Also I tested.  We maybe have to add a `#define` with the guard of the header of  ` std::iterator`. Or do we have a testsuite where the deprecated no longer exists.

Comment (lrineau) [reply STL_Extension/include/CGAL/iterator.h]: - `CGAL::iterator_traits` does not exist, does it?
- `std::iterator_traits` is not deprecated.

Comment (afabri) [reply STL_Extension/include/CGAL/iterator.h]: you are right.

**Comment by lrineau** [comment]
Tested successfully in [CGAL-4.12-Ic-182], and then merged.

[CGAL-4.12-Ic-182]: https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-182.shtml

**Pull request merged**